reactor-core/src/main/java/reactor/core/publisher/FluxSwitchMapNoPrefetch.java

```diff
@@ -0,0 +1,900 @@
+			this.actual.onError(Exceptions.terminate(THROWABLE, this));
+		}
+
+		@Override
+		public void onComplete() {
+			if (this.done) {
+				return;
+			}
+
+			this.done = true;
+
+			final long state = setMainCompleted(this);
+			if (state == TERMINATED) {
+				return;
+			}
+
+			final SwitchMapInner<T, R> inner = this.inner;
+			if (inner == null || hasInnerCompleted(state)) {
+				this.actual.onComplete();
+			}
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				long previousRequested = Operators.addCap(REQUESTED, this, n);
+				long state = addRequest(this, previousRequested);
+
+				if (state == TERMINATED) {
+					return;
+				}
+
+				if (hasRequest(state) == 1 && isInnerSubscribed(state) && !hasInnerCompleted(state)) {
+					final SwitchMapInner<T, R> inner = this.inner;
+					if (inner.index == index(state)) {
+						inner.request(n);
+					}
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			final long state = setTerminated(this);
+			if (state == TERMINATED) {
+				return;
+			}
+
+			final SwitchMapInner<T, R> inner = this.inner;
+			if (inner != null && isInnerSubscribed(state) && !hasInnerCompleted(state) && inner.index == index(state)) {
+				inner.cancel();
+			}
+
+			if (!hasMainCompleted(state)) {
+				this.s.cancel();
+
+				//ensure that onError which races with cancel has its error dropped
+				final Throwable e = Exceptions.terminate(THROWABLE, this);
+				if (e != null) {
+					Operators.onErrorDropped(e, this.actual.currentContext());
+				}
+			}
+		}
+	}
+
+	static final class SwitchMapInner<T, R> implements InnerConsumer<R> {
+
+		final SwitchMapMain<T, R> parent;
+		final CoreSubscriber<? super R> actual;
+
+		final int index;
+
+		Subscription s;
+		long         produced;
+		long         requested;
+		boolean      done;
+
+		T nextElement;
+		SwitchMapInner<T, R> nextInner;
+
+		SwitchMapInner(SwitchMapMain<T, R> parent, CoreSubscriber<? super R> actual, int index) {
+			this.parent = parent;
+			this.actual = actual;
+			this.index = index;
+		}
+
+		@Override
+		public Context currentContext() {
+			return parent.currentContext();
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			if (key == Attr.CANCELLED) return this.index ;
+			if (key == Attr.PARENT) return this.parent;
+			if (key == Attr.ACTUAL) return this.actual;
+			if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+
+			return null;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(this.s, s)) {
+				this.s = s;
+
+				final int expectedIndex = this.index;
+				final SwitchMapMain<T, R> parent = this.parent;
+				final long state = setInnerSubscribed(parent, expectedIndex);
+
+				if (state == TERMINATED) {
+					s.cancel();
+					return;
+				}
+
+				final int actualIndex = index(state);
+				if (expectedIndex != actualIndex) {
+					s.cancel();
+					parent.subscribeInner(this.nextElement, this.nextInner, actualIndex);
```

Comment:
wouldn't that be done inside the `onSubscribe` of the next inner?

Reply:
I designed the process in such a way, so subscribeInner is not thread-safe and must be performed serially. Having a state machine there, I guarantee that if we have subscribed to inner, but inner is still sending a subscription, the racing onNext will just increment index without trying to call `subscribeInner`, thus once inner onSubscribe happens, it will try to move state machine and set flag inner subscribed but fail since this inner index will not much the index set in the volatile state, hence, it will cancel the current subscription and subscribe to the next inner which was set with happens-before relationship. Same for Inner#OnNext or Inner#onComplete, if at some point of touching volatile state we see that index has changed (Inner#index != index(state)), we call subscribeNext (if we were under WIP) and call subscribeInner). The only case when subscribeInner is called from main#OnNext is when the ongoing inner subscription has subscribed and has no WIP flag set (which means there will not be any racing between the next inner and the current inner and by that we guarantee that we will never have actual#onXXX racing)

Reply:
got it